service/controller/release/resource/apps/resource_test.go

```diff
@@ -0,0 +1,172 @@
+					key.ConstructApp(testComponents[2]),
+				},
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Log(tc.name)
+
+			resultApps := calculateMissingApps(tc.operators, tc.apps)
+
+			if !cmp.Equal(resultApps, tc.expectedApps) {
+				t.Fatalf("\n\n%s\n", cmp.Diff(tc.expectedApps, resultApps))
+			}
+		})
+	}
+}
+
+func Test_calculateObsoleteApps(t *testing.T) {
+	testCases := []struct {
+		name         string
+		operators    map[string]releasev1alpha1.ReleaseSpecComponent
+		apps         appv1alpha1.AppList
+		expectedApps appv1alpha1.AppList
+	}{
+		{
+			name: "case 0: there is an obsolete app",
+			operators: map[string]releasev1alpha1.ReleaseSpecComponent{
+				key.BuildAppName(testComponents[0]): testComponents[0],
+				key.BuildAppName(testComponents[1]): testComponents[1],
+			},
+			apps: appv1alpha1.AppList{
+				Items: []appv1alpha1.App{
+					key.ConstructApp(testComponents[0]),
+					key.ConstructApp(testComponents[1]),
+					key.ConstructApp(testComponents[2]),
+				},
+			},
+			expectedApps: appv1alpha1.AppList{
+				Items: []appv1alpha1.App{
+					key.ConstructApp(testComponents[2]),
+				},
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Log(tc.name)
+
+			resultApps := calculateObsoleteApps(tc.operators, tc.apps)
+
+			if !cmp.Equal(resultApps, tc.expectedApps) {
+				t.Fatalf("\n\n%s\n", cmp.Diff(tc.expectedApps, resultApps))
+			}
+		})
+	}
+}
+
+func Test_excludeDeletedRelease(t *testing.T) {
+	testCases := []struct {
+		name             string
+		releases         releasev1alpha1.ReleaseList
+		expectedReleases releasev1alpha1.ReleaseList
+	}{
+		{
+			name: "case 0: some releases are being deleted",
+			releases: releasev1alpha1.ReleaseList{
+				Items: []releasev1alpha1.Release{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:              "being-deleted",
+							DeletionTimestamp: &metav1.Time{},
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:              "not-being-deleted",
+							DeletionTimestamp: nil,
+						},
+					},
+				},
+			},
+			expectedReleases: releasev1alpha1.ReleaseList{
+				Items: []releasev1alpha1.Release{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "not-being-deleted",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Log(tc.name)
+
+			resultReleases := excludeDeletedRelease(tc.releases)
+
+			if !cmp.Equal(resultReleases, tc.expectedReleases) {
+				t.Fatalf("\n\n%s\n", cmp.Diff(tc.expectedReleases, resultReleases))
+			}
+		})
+	}
+}
+
+func appForComponent(operator releasev1alpha1.ReleaseSpecComponent) appv1alpha1.App {
+	return appv1alpha1.App{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: key.BuildAppName(operator),
+		},
+		Spec: appv1alpha1.AppSpec{
+			Name: operator.Name,
+			//TODO this should be ref
```

Comment:
What's the plan for this TODO?

Reply:
Need to double check - I think the version is correct after all. Maybe you know @yulianedyalkova

Reply:
The idea here was that if the test operators had a reference, that will be the value being used here instead of the version. But since all of the decisions if version or reference is used went to the key.go file, that's not needed here anymore :)

Reply:
Okay, want to just drop this `TODO` then?